modules/talkmuted/TalkMutedDetection.js

```diff
@@ -0,0 +1,57 @@
+import * as JitsiTrackEvents from '../../JitsiTrackEvents';
+
+export default class TalkMutedDetection {
+    /**
+     * Creates TalkMutedDetection
+     * @param callback the callback to call when detected local user is talking
+     * while its microphone is muted.
+     * @constructor
+     */
+    constructor(callback) {
+        this.callback = callback;
+
+        // we track firing the event, in order to avoid sending too many events
+        this.eventFired = false;
+    }
+
+    /**
+     * Receives audio level events for all send/receive streams.
+     * @param ssrc the ssrc of the stream
+     * @param level the current audio level
+     * @param isLocal whether this is local or remote stream (sent or received)
+     */
+    audioLevelListener(ssrc, level, isLocal) {
+        // we are interested only in local audio stream
+        // and if event is not already sent
+        if (!isLocal || !this.audioTrack || this.eventFired)
+            return;
+
+        if (this.audioTrack.isMuted() && level > 0.6) {
+            this.eventFired = true;
+            this.callback();
+        }
+    }
+
+    /**
+     * Mute changed for a track.
+     * @param track the track which mute state has changed.
+     */
+    muteChanged(track) {
+        if (!track.isLocal() || !track.isAudioTrack())
+            return;
+
+        if (track.isMuted())
+            this.eventFired = false;
```

Comment:
1. Obviously, the total number of lines above can be reduced:

```` javascript
if (track.isLocal() && track.isAudioTrack() && track.isMuted())
    this.eventFired = false;
````

Since you haven't done that, I'm thinking that you have a reason. The only reason that I see is that you want to make it clear, explicit that regardless of what `muteChanged` does, it will not do it for non-local, non-audio tracks. But then isn't that a condition for the whole `TalkMutedDetection`? If it is, why isn't `addTrack` using the same condition and it's using just a part of it instead? I find such differences to be adding complexity because one has to go and look at the method call sites to determine that indeed the condition will be met.